src/lib/swapper/swappers/OneInchSwapper/approvalNeeded/approvalNeeded.ts

```diff
@@ -16,48 +18,50 @@ import type {
 export const approvalNeeded = async (
   deps: OneInchSwapperDeps,
   input: ApprovalNeededInput<EvmChainId>,
-): Promise<ApprovalNeededOutput> => {
+): Promise<Result<ApprovalNeededOutput, SwapErrorRight>> => {
   const { accountNumber, allowanceContract, sellAmountBeforeFeesCryptoBaseUnit, sellAsset } =
     input.quote
 
   const adapterManager = getChainAdapterManager()
   const adapter = adapterManager.get(sellAsset.chainId)
 
   if (adapter === undefined) {
-    throw new SwapError('[approvalNeeded] - getChainAdapterManager returned undefined', {
-      code: SwapErrorType.UNSUPPORTED_NAMESPACE,
-      details: { chainId: sellAsset.chainId },
-    })
+    return Err(
+      makeSwapErrorRight({
+        message: '[approvalNeeded] - getChainAdapterManager returned undefined',
+        code: SwapErrorType.UNSUPPORTED_NAMESPACE,
+        details: { chainId: sellAsset.chainId },
+      }),
+    )
   }
 
   const { assetReference: sellAssetErc20Address } = fromAssetId(sellAsset.assetId)
 
-  try {
-    const receiveAddress = await adapter.getAddress({ accountNumber, wallet: input.wallet })
+  const receiveAddress = await adapter.getAddress({ accountNumber, wallet: input.wallet })
 
-    if (!allowanceContract) {
-      throw new SwapError('[approvalNeeded] - allowanceTarget is required', {
+  if (!allowanceContract) {
+    return Err(
+      makeSwapErrorRight({
+        message: '[approvalNeeded] - allowanceTarget is required',
         code: SwapErrorType.VALIDATION_FAILED,
         details: { chainId: sellAsset.chainId },
-      })
-    }
-
-    const apiInput: OneInchAllowanceApiInput = {
-      tokenAddress: sellAssetErc20Address,
-      walletAddress: receiveAddress,
-    }
-    const { chainReference } = fromChainId(sellAsset.chainId)
-
-    const allowanceResponse: AxiosResponse<OneInchAllowanceResponse> = await axios.get(
-      `${deps.apiUrl}/${chainReference}/approve/allowance`,
-      { params: apiInput },
+      }),
     )
-    const allowanceOnChain = bnOrZero(allowanceResponse.data.allowance)
+  }
 
-    return {
-      approvalNeeded: allowanceOnChain.lt(bnOrZero(sellAmountBeforeFeesCryptoBaseUnit)),
-    }
-  } catch (e) {
-    throw new SwapError('[approvalNeeded]', { cause: e, code: SwapErrorType.CHECK_APPROVAL_FAILED })
+  const apiInput: OneInchAllowanceApiInput = {
+    tokenAddress: sellAssetErc20Address,
+    walletAddress: receiveAddress,
   }
+  const { chainReference } = fromChainId(sellAsset.chainId)
+
+  const allowanceResponse: AxiosResponse<OneInchAllowanceResponse> = await axios.get(
+    `${deps.apiUrl}/${chainReference}/approve/allowance`,
+    { params: apiInput },
+  )
+  const allowanceOnChain = bnOrZero(allowanceResponse.data.allowance)
```

Comment:
I feel like we need an optional chain here in case the request fails. Maybe it just throws though, so it's fine.

```suggestion
  const allowanceOnChain = bnOrZero(allowanceResponse.data?.allowance)
```

Reply:
The naive answer I wrote was that I would trust the types here - `data` should actually always been defined.

In reality, as a general pattern, it may not be truthy (i.e `{data: undefined}` which is perfectly valid, although looking at the docs this isn't the case here) which is precisely what the monadic axios services are checking for.

Though case on point `OneInchSwapper` doesn't take the same approach of using a `oneInchService` which is problematic, since the services are now proxied monadic axios instances, which *do* check for the falsy value of `response.data`.

Might as well then: a45340d5415b9d8084ee8bb757678f5a9dfad3db cc @0xean, FYI I honored the original implementation here, but we may or may not be able to cache URLs here.